Add support for image digests in helm charts

aanm opened the pull request:
Read per commit

```release-note
Add digest flags to specify docker images digests in helm charts
```

aanm committed:
install: add digests into helm charts

This commit adds a simple way to add the image digests into the official
helm charts.

Signed-off-by: André Martins <andre@cilium.io>

aanm committed:
images/scripts: fix get-image-digest to work on any image

Signed-off-by: André Martins <andre@cilium.io>

aanm committed:
contrib/release: do not require images to be download locally

The SHAs are available in the GH run, so we can use the GH API to
retrieve them automatically.

Signed-off-by: André Martins <andre@cilium.io>

aanm committed:
.github: generate release files automatically

Having files with the right format will make it easier to download them
locally and use them as part of the release process.

Signed-off-by: André Martins <andre@cilium.io>

aanm committed:
.github/ISSUE_TEMPLATE/release_template.md

re-write release process to have more automation

Signed-off-by: André Martins <andre@cilium.io>

christarazi commented on install/kubernetes/cilium/templates/cilium-agent-daemonset.yaml:
```diff
@@ -239,7 +239,11 @@ spec:
 {{- end }}
+{{- if .Values.image.useDigest }}
+        image: {{ .Values.image.repository }}:{{ .Values.image.tag }}@{{ .Values.image.digest }}
+{{- else }}
```
Just wondering, could we eliminate `useDigest` and simply rely on `len(digest) != 0`? I think `empty` would work: https://helm.sh/docs/chart_template_guide/function_list/#empty

My thinking is that adding 1 new variable is better than 2.

aanm replied:
right but there might be users that don't want to use the SHA for some reason, that's why there are 2 variables instead of 1.

nbusseneau replied:
I'm thinking we could just rely `digest` itself being set or not, rather than have a separate `useDigest` boolean. For example with:
```yaml
{{- if .Values.image.digest }}
        image: {{ .Values.image.repository }}:{{ .Values.image.tag }}@{{ .Values.image.digest }}
{{- else }}
```
If `digest` is `""` as in the `README` examples then it won't enter the `if` block.

aanm replied:
@Skymirrh the digests will be part of the helm charts but using them will be optional. For example our CI will not use the image digests, even tho they are defined, so the digests will need to be set false when setting up our CI (once we branch off to 1.10)

nbusseneau replied:
I agree, but even then, doesn't it make sense to just programmatically remove `digest` or set `digest: ""` to clear digest usage in CI, rather than set `useDigest: false`? What I'm trying to say is that since we already have a manual action needed for setting up values for CI anyway, why not just remove complexity by having a single value instead of a separate flag?

aanm replied:
I think I got it now, yes that makes sense. However, the PRs on the stable branches will need to have the `useDigest` because we want to have the digests available but they should not be mandatory to users, as they it would be a breaking change.

nbusseneau replied:
I'm a bit confused: could you detail in which scenario it is possible to have `digest: "{someSha}"` but at the same time not wanting to use it?

I was under the assumption that we could limit ourselves to these two scenarios:
- `digest` not present or `digest: ""` > current behavior (no breaking change)
- `digest` present > new behavior

aanm replied:
@Skymirrh in stable branches, not master we don't want to introduce this new behavior.

nbusseneau replied:
Ah, so if I understood correctly, in stable branches, we will have:
```yaml
image:
    digest: "{someSha}"
    useDigest: false
```
Because we want users to be able to easily switch `useDigest` to `true` and use some predefined digests. Correct?

aanm replied:
@Skymirrh correct.

nbusseneau replied:
OK, works for me! Then a small nitpick proposal, in case you like it: instead of
```yaml
{{- if .Values.image.useDigest }}
        image: {{ .Values.image.repository }}:{{ .Values.image.tag }}@{{ .Values.image.digest }}
{{- else }}
        image: {{ .Values.image.repository }}:{{ .Values.image.tag }}
{{- end }}
```
we could have
```yaml
        image: {{ .Values.image.repository }}:{{ .Values.image.tag }}{{ if .Values.image.useDigest }}@{{ .Values.image.digest }}{{ end }}
```
I tend to use the latter construct when it's a simple flag switch added to an existing block, rather than a wholly different block.

@christarazi feel free to resolve when/if you are satisfied :)

aanm replied:
@Skymirrh I gave it a little more thought over the weekend. I think it's better to leave the `useDigest` mostly for the cilium-operator where we have 3 different image versions. Picking the right image to run in the right cluster is hidden from the user, i.e., the image chosen depends on the options users set. Since this behavior is hidden, users would need to be aware which digest (`awsDigest`, `azureDigest`, etc) they would need to disable. Let me know what do you think.

nbusseneau replied:
Makes sense. I wonder to what extent we should document this.

christarazi commented on images/scripts/get-image-digest.sh:
```diff
@@ -12,4 +12,10 @@ if [ "$#" -ne 1 ] ; then
 fi
 
-docker buildx imagetools inspect "${1}" 2>/dev/null | grep Digest | awk '{ print $2 }' || true
+inspect=$(docker buildx imagetools inspect "${1}" --raw 2>/dev/null | sha256sum | cut -d " " -f 1 )
```
This is not meant to run on the developer's machine ever, right? From what I remember, Darwin's hash functions are invoked differently than Linux

aanm replied:
oh I didn't know that... what output do you get on mac os for let's say this:
```
$ ./images/scripts/get-image-digest.sh docker.io/cilium/cilium:v1.7.14
sha256:0bd894e9dc0188482bc0316c2889a0aa9b2fce3a3c9e4dc6f55afb4a2afbf84b
```

christarazi replied:
I don't have Mac OS, but I remember this from when I used to use it in a past life :wink:

aditighag replied:
> what output do you get on mac os for let's say this:
> ```
> $ ./images/scripts/get-image-digest.sh docker.io/cilium/cilium:v1.7.14
> sha256:0bd894e9dc0188482bc0316c2889a0aa9b2fce3a3c9e4dc6f55afb4a2afbf84b
> ```

@aanm It doesn't print anything.

aanm replied:
@aditighag can you try again please?

christarazi replied:
As long as the script is intended for GH Actions, then maybe we don't need worry about this

aditighag replied:
> @aditighag can you try again please?

@aanm ~The command [1] itself works, but the script doesn't print anything. I narrowed it down to `set -o errexit` *and* `set -o pipefail`. I'll need to look into this further~.

Edit: The docker command syntax is different on mac OS  (or I'm missing some packages) - 

```
docker buildx imagetools inspect "docker.io/cilium/cilium:v1.7.14"
docker: 'buildx' is not a docker command.
See 'docker --help'
```

Edit2: I upgraded Docker on my laptop, and it works now. 🚀

```
./images/scripts/get-image-digest.sh docker.io/cilium/cilium:v1.7.14
sha256:0bd894e9dc0188482bc0316c2889a0aa9b2fce3a3c9e4dc6f55afb4a2afbf84b
```

[1] `docker buildx imagetools inspect "${1}" --raw 2>/dev/null | openssl sha256 -r | cut -d " " -f 1`

christarazi commented on .github/workflows/images-legacy-releases.yaml:
```diff
@@ -114,4 +123,30 @@ jobs:
+        run: |
+          cd image-digest/
+          echo "# File generated by .github/workflows/images-legacy-releases.yaml; DO NOT EDIT." > ../Makefile.digests
+          echo "# Copyright 2021 Authors of Cilium" >> ../Makefile.digests
```
We could use `date +'%Y'` to automate fetching the year :wink:

christarazi commented on contrib/release/pull-docker-manifests.sh:
```diff
@@ -27,39 +31,55 @@ handle_args() {
+get_digest_output() {
+    local username run_id file tmp_dir archive_download_url archive_download_url_zip
+
+    username=${1}
```
Nit: quotes?

christarazi commented on contrib/release/pull-docker-manifests.sh:
```diff
@@ -3,21 +3,25 @@
+    logecho "usage: $0 <GH-USERNAME> <VERSION> <RUN-URL>"
+    logecho "GH-USERNAME  GitHub username"
+    logecho "VERSION      Target version"
+    logecho "RUN-URL      GitHub URL with the RUN for the release images"
```
I think what you mean here is `RUN-ID` rather than the full URL, just looking at how the variable is used. It would not work if the full URL was passed. Also, I think we can be clearer on which job specifically has this ID. Could you put the job name in the help string so that it is unambiguous?

christarazi replied:
I saw in later commits that we are supposed to save the URL from the GH run, but I think it would still help to put slightly more clarification.

aanm replied:
No, I meant `RUN-URL`, see line 69 where the `run_id` is derived from the url

christarazi replied:
TIL. I didn't know `basename` can work with URLs :exploding_head:

christarazi commented on .github/ISSUE_TEMPLATE/release_template.md:
```diff
@@ -33,24 +33,27 @@ assignees: ''
     `make -C install/kubernetes/ check-docker-images`
-- [ ] Get the image digests from the build process (click in the GH Run > "Display Digests"),
-      update the digests in the helm charts locally and make a commit / PR with the digests.
+- [ ] Get the image digests from the build process make a commit / PR with the digests.
```
```suggestion
- [ ] Get the image digests from the build process and make a commit / PR with the digests.
```

christarazi commented on .github/ISSUE_TEMPLATE/release_template.md:
```diff
@@ -33,24 +33,27 @@ assignees: ''
   - [ ] Update the text at the top with 2-3 highlights of the release
-  - [ ] Run `contrib/release/pull-docker-manifests.sh` to fetch the image SHAs
-        and copy the text into the end of the release
+  - [ ] Copy the text from `digest-vX.Y.0.txt` (previously generated with
```
It is always `v.X.Y.**0**`?

aanm replied:
no, is suppose to be `X.Y.Z`

christarazi reviewed (commented):
Nice! Some comments to clarify and address mostly minor things

aditighag commented on images/scripts/get-image-digest.sh:
```diff
@@ -12,4 +12,16 @@ if [ "$#" -ne 1 ] ; then
-docker buildx imagetools inspect "${1}" 2>/dev/null | grep Digest | awk '{ print $2 }' || true
+if [ "$(uname)" != "Linux" ]; then
+  # shellcheck disable=SC2120
+  function sha256sum() { openssl sha256 "$@" | awk '{print $2}'; }
```
```suggestion
  function sha256sum() { openssl sha256; }
```

This doesn't work on mac OS. I verified that the suggested change works.

aditighag reviewed (approved):
I only reviewed `images/scripts/get-image-digest.sh` changes.

joestringer commented on contrib/release/pull-docker-manifests.sh:
```diff
@@ -27,39 +31,55 @@ handle_args() {
+get_digest_output() {
+    local username run_id file tmp_dir archive_download_url archive_download_url_zip
+
+    username="${1}"
```
We have other scripts autodetecting the username these days so we can likely remove that parameter, but that's a super minor nit.

joestringer commented on .github/ISSUE_TEMPLATE/release_template.md:
```diff
@@ -33,24 +33,28 @@ assignees: ''
       [cilium helm release tool] for both the `vX.Y.Z` release and `vX.Y` branch
-      & push to repository
+      and push these changes into the helm repository as well as make a commit
+      and push them into cilium repository.
```
nit: At the moment it's not *so* obvious that the pull-docker-manifests is actually generating different files in the tree and those are the changes that need to be created into the new commit.

Same content, minor reorganisation.

```suggestion
- [ ] Update the Helm charts in the [Cilium charts] repository and the main repository
  - [ ] Run `contrib/release/pull-docker-manifests.sh` in the Cilium repository to fetch the image digests, use the URL of the GitHub run here.
  - [ ] Run the [cilium helm release tool] from the cilium-charts repository to generate the `vX.Y.Z` release charts and `vX.Y` branch charts.
  - [ ] Push the new helm charts to the Cilium charts repository.
  - [ ] After pushing the charts, add the local changes from the Cilium repository to a new commit and open a PR against the release branch to update the digests and charts.
```

aanm replied:
Discussed offline, it should be something around:
```
- [ ] Commit and merge into cilium/cilium
- [ ] Fetch the branch
- [ ] Create helm charts artifacts in [Cilium charts] repository using
      [cilium helm release tool] for both the `vX.Y.Z` release and `vX.Y` branch
      and push these changes into the helm repository.
```

joestringer reviewed (approved):
Two minor nits, otherwise LGTM.

I agree it'd be nice to not have the `useDigest` flag but for the older branches we need that to avoid overriding user configuration since we didn't make those flags available in the first place.

aanm committed:
.github: remove comments of GH action versions

Dependabot will update those commit SHAs automatically and the version
number will quickly outdated. To avoid confusion, we will remove the
comments related with the versions a SHA points to from all GH actions.

Signed-off-by: André Martins <andre@cilium.io>
